jobs/job_test_utils.py

```diff
@@ -44,30 +48,91 @@ def __init__(self, *args, **kwargs):
         self.pipeline = test_pipeline.TestPipeline(
             runner=runners.DirectRunner(),
             options=test_pipeline.PipelineOptions(runtime_type_check=True))
+        self._close_stack = None
 
-    def run(self, *args, **kwargs):
-        """Runs the test within the context of a test pipeline."""
-        with self.pipeline:
-            super(PipelinedTestBase, self).run(*args, **kwargs)
+    def setUp(self):
+        super(PipelinedTestBase, self).setUp()
+        with contextlib2.ExitStack() as stack:
+            stack.enter_context(decorate_beam_errors())
+            stack.enter_context(self.pipeline)
+            self._close_stack = stack.pop_all().close
+
+    def tearDown(self):
+        try:
+            self.flush_pipeline(strict=False)
+        finally:
+            super(PipelinedTestBase, self).tearDown()
+
+    def flush_pipeline(self, strict=True):
+        """Flushes the pipeline wrapping the test to end immediately.
+
+        Args:
+            strict: bool. Whether to raise an exception when the pipeline has
+                been flushed twice.
+
+        Raises:
+            RuntimeError. The pipeline has been closed twice and strict is True.
+        """
+        if self._close_stack:
+            self._close_stack()
+            self._close_stack = None
+        elif strict:
+            raise RuntimeError('flush_pipeline() must be called exactly once')
 
     def assert_pcoll_equal(self, actual, expected):
         """Asserts that the given PCollections are equal.
 
+        NOTE: Only one assert_pcoll_*() method may be called in a test.
```

Comment:
While this addresses the "assert is not an assert" issue, this is quite restrictive. Is it not possible to do more at a time, e.g. reset the pipeline instead of closing it?

Reply:
I don't think that'd be a good idea, since once the pipeline is flushed the data is gone with it; they'd have to rebuild the entire pipeline from scratch.

That feels too unintuitive and inefficient (especially considering it'd be equivalent to writing a new test case).

Reply:
OK, fair enough. Please add all this reasoning to the docstring of why we can't relax this and why we're doing it this way. We can revisit if the need arises for it, but I think it's fine to live with this restriction for now. Thanks!